### PR TITLE
MAYA-128420 Import Chaser doesn't recieve information on how to match USD prim to DAG paths

### PR DESCRIPTION
**Pull request by samuelliu-adsk**

### CODE REVIEW

**Review by vlasovi — CHANGES_REQUESTED** [review]
Overall looks good, but I think we may need to do some improvements:
1. I noticed that you have two different code paths needed for C++ case and for Python. One stores the map in UsdMayaImportChaser and the other in FactoryContext. I think we need to choose only one and go with it.
2. Also you use two different maps: sdfToDagMap and newRegistry. They seem pretty similar to me. Can one map cover all our needs?
3. Also, I would rename WriteToNodeRegistry to SetNodeRegisrty, as this function doesn't really write anything

**Comment by samuelliu-adsk** [comment]
@vlasovi 
Thanks for the comments. I forgot to remove the nodeRegistry, that was a previous design. Thanks!

**Comment by samuelliu-adsk** [comment]
> 1. I noticed that you have two different code paths needed for C++ case and for Python. One stores the map in UsdMayaImportChaser and the other in FactoryContext. I think we need to choose only one and go with it.

Hi @vlasovi ,

It's a really good point but I don't see a way to just keep one.. 
The "sdfToDagMap" in ImportChaser is for cpp chaser, which will inherit the map from UsdMayaImportChaser class. However, the one in factory context is for python only. To pass reference to python using boost library, we need a member variable in the class, so we have to keep one inside our factoryContext class.

Correct me if I'm wrong or there's a potential work around I didn't think of!

Thanks!
Samuel

File: lib/mayaUsd/fileio/chaser/importChaser.h [hunk 1]
```diff
@@ -18,6 +18,7 @@
 
 #include <mayaUsd/base/api.h>
 #include <mayaUsd/fileio/jobs/jobArgs.h>
+#include <mayaUsd/fileio/primReaderContext.h>
```
Comment (vlasovi) [inline lib/mayaUsd/fileio/chaser/importChaser.h:21]: I think this include is not needed

**Comment by vlasovi** [comment]
@samuelliu-adsk One way to go about accessing the map is to have function GetSdfToDagMap exported to python not from FactoryContext class but form ImportChaser class. This way you don't need to store the map in two places.

File: lib/mayaUsd/fileio/jobs/readJob.cpp [hunk 2]
```diff
@@ -428,6 +428,23 @@ bool UsdMaya_ReadJob::Read(std::vector<MDagPath>* addedDagPaths)
...
     }
 
+    SdfPathSet   sdfImportedPaths;
+    MSdfToDagMap sdfToDagMap;
+    for (const UsdPrim prim : stage->TraverseAll()) {
+        sdfImportedPaths.insert(prim.GetPath());
+    }
+
+    TF_FOR_ALL(pathsIter, sdfImportedPaths)
+    {
```
Comment (vlasovi) [inline lib/mayaUsd/fileio/jobs/readJob.cpp:438]: I don't understand why you need two cycles here. You don't need to accumulate sdfImportedPaths first and then loop over it. You can what you need inside the first  loop.

File: lib/mayaUsd/fileio/jobs/readJob.cpp [hunk 3]
```diff
@@ -437,16 +454,20 @@ bool UsdMaya_ReadJob::Read(std::vector<MDagPath>* addedDagPaths)
                 addedDagPaths->push_back(MDagPath::getAPathTo(obj));
                 currentAddedDagPaths.append(MDagPath::getAPathTo(obj));
                 fromSdfPaths.push_back(pathsIter->GetPrimPath());
+                sdfToDagMap[pathsIter->GetPrimPath()] = MDagPath::getAPathTo(obj);
```
Comment (vlasovi) [inline lib/mayaUsd/fileio/jobs/readJob.cpp:457]: Why do you need to do this here? Wasn't everything already collected in the loop above?

File: test/lib/mayaUsd/fileio/testImportChaser.py [hunk 4]
```diff
@@ -67,6 +67,24 @@ def PostImport(self, returnPredicate, stage, dagPaths, sdfPaths, jobArgs):
...
+        for sdfToDag in sdfToDagMap:
+            allSdfPaths.append(sdfToDag.key())
+            allDagPaths.append(sdfToDag.data())
+
+        assert(len(allSdfPaths) == len(allDagPaths))
+
+        for i in range(len(allSdfPaths)):
+            print(("----------primitive ") + str(i))
+            print(str(allSdfPaths[i]))
+            print(str(allDagPaths[i]))
```
Comment (vlasovi) [inline test/lib/mayaUsd/fileio/testImportChaser.py:86]: I don't think you are actually testing anything here. You need to have the code that asserts, and not just printing. You can actually remove printing as this isn't very useful.

File: lib/mayaUsd/fileio/jobs/readJob.cpp (hunk 2 above)
Comment (samuelliu-adsk) [reply lib/mayaUsd/fileio/jobs/readJob.cpp:438]: I forgot to delete it...Thanks for pointing out

**Comment by samuelliu-adsk** [comment]
> @samuelliu-adsk One way to go about accessing the map is to have function GetSdfToDagMap exported to python not from FactoryContext class but form ImportChaser class. This way you don't need to store the map in two places.

Hi @vlasovi ,

I tried some approaches but I don't think we can do that without changing the whole code structure. To call that "GetSdfToDagMap" function, we need to pass the reference of the class to python. Currently we only pass factoryContext to python, and factory context does not share any hierarchy relationship with MayaUsdImportChaser, so we can't use get the variables from ImportChaser in FactoryContext scope (i.e: In python script we need to call FactoryContext.GetJobArgs() ). That's why other variables (sdfPrimPaths and etc.) are passed into factory context as well.

Thanks,
Samuel

**Comment by vlasovi** [comment]
@samuelliu-adsk importChaserTest python class inherits from mayaUsdLib.ImportChaser python which in it's turn originates from ImportChaserWrapper c++ class, which inherits from UsdMayaImportChaser c++ class, so I think you should be able to access any methods of UsdMayaImportChaser inside importChaserTest

**Review by vlasovi — CHANGES_REQUESTED** [review]
Looks very good now. Only couple more things to finish:
1. In testImportChaser.py you need to add asserts to autotest the new feature. I would also remove prints as they may be polluting too much the autotest log. 
2. Rename WriteToSdfToDagMap to SetSdfToDagMap and also use const& for the parameter to make sure there is no extra copying of the potentially large map.

File: test/lib/mayaUsd/fileio/testImportChaser.py [hunk 5]
```diff
@@ -67,6 +67,24 @@ def PostImport(self, returnPredicate, stage, dagPaths, sdfPaths, jobArgs):
...
```
Comment (seando-adsk) [inline test/lib/mayaUsd/fileio/testImportChaser.py:86]: Why is the test printing stuff - is that just for you to debug?

**Review by seando-adsk — CHANGES_REQUESTED** [review]

File: test/lib/mayaUsd/fileio/testImportChaser.py (hunk 5 above)
Comment (samuelliu-adsk) [reply test/lib/mayaUsd/fileio/testImportChaser.py:86]: No actually I want to do a demo here to show the customers how to use the feature..

Comment (samuelliu-adsk) [reply test/lib/mayaUsd/fileio/testImportChaser.py:86]: If it's not needed I can remove it

File: test/lib/mayaUsd/fileio/testImportChaser.py [hunk 6]
```diff
@@ -55,6 +58,18 @@ def PostImport(self, returnPredicate, stage, dagPaths, sdfPaths, jobArgs):
...
+        # Get the Sdf path to Dag path matching of all imported prims
+        # Call GetSdfToDagMap() to get the sdfToDagMap, 
+        # which contains SDF path to Dag path mapping
+        sdfToDagMap = self.GetSdfToDagMap()
+        assert(primCount == len(sdfToDagMap))
+        sdfToDagMappingStr = "Sdf to Dag mapping: \n"
+        for sdfToDag in sdfToDagMap:
+            sdfToDagMappingStr += (str(sdfToDag.key()) + "\n")
+            sdfToDagMappingStr += (str(sdfToDag.data()) + "\n")
+
```
Comment (vlasovi) [inline test/lib/mayaUsd/fileio/testImportChaser.py:70]: One more thing you should do here is to verify the content of sdfToDagMap like: 
self.assertEqual(sdfToDagMap[0].key(), "WhateverKeyExpect");
self.assertEqual(sdfToDagMap[0].data(), "WhateverDataExpect");
....

**Review by vlasovi — CHANGES_REQUESTED** [review]
Please verify the content of sdfToDagMap in the autotest

**Review by vlasovi — APPROVED** [review]

File: test/lib/mayaUsd/fileio/testImportChaser.py (hunk 6 above)
Comment (samuelliu-adsk) [reply test/lib/mayaUsd/fileio/testImportChaser.py:76]: Thanks for the suggestion!

**Review by vlasovi — APPROVED** [review]

**Review by seando-adsk — APPROVED** [review]
Looks good Samuel. Great work.

**Pull request merged**